File: Top Interview Questions/L130. 被围绕的区域.ts
/**
 Do not return anything, modify board in-place instead.
 */
function solve(board: string[][]): void {
    let x = board.length
    let y = board[0].length
    for (let i = 0; i < x; i++) {
        if (board[i][0] == 'O') {
            dfs(board, i, 0)
        }
    }
    for(let i=0;i<x;i++){
        if(board[i][y-1]=='O'){
            dfs(board,i,y-1)
        }
    }
    for(let i=0;i<y;i++){
        if(board[0][i]=='O'){
            dfs(board,0,i)
        }
    }
    for(let i=0;i<y;i++){
        if(board[x-1][i]=='O'){
            dfs(board,x-1,i)
        }
    }

    for(let i=0;i<x;i++){
        for(let j=0;j<y;j++){
            if(board[i][j]=='-'){
                board[i][j]='O'
            }else{
                board[i][j]='X'
            }
        }
    }

    function dfs(board: string[][], x: number, y: number) {
        board[x][y] = '-'
        if (x - 1 > 0 && board[x - 1][y] == 'O') {
            // board[x - 1][y] = '-'
            dfs(board, x - 1, y)
        }
        if (x + 1 < board.length && board[x + 1][y] == 'O') {
            // board[x + 1][y] = '-'
            dfs(board, x + 1, y)
        }
        if (y - 1 > 0 && board[x][y - 1] == 'O') {
            // board[x][y - 1] == '-'
            dfs(board, x, y - 1)
        }
        if (y + 1 < board[0].length && board[x][y + 1] == 'O') {
            // board[x][y + 1] == '-'
            dfs(board, x, y + 1)
        }
    }
};

File: golang-devildyw/middle/L130被围绕的区域.go
package main

func main() {}

func solve(board [][]byte) {
	//既然在的'O'不算被包围的 那么我们从边缘起手找到那些'O'并且找出与那些'O' 相连的'O'并把他们标记为'F'
	if board == nil || len(board) == 0 {
		return
	}

	m := len(board)
	n := len(board[0])

	for i := 0; i < m; i++ {
		// 边缘按行遍历 边缘即位行的两边
		dfs1(board, i, 0)
		dfs1(board, i, n-1)
	}

	for j := 0; j < n; j++ {
		//按列遍历，边缘即位列的两边
		dfs1(board, 0, j)
		dfs1(board, m-1, j)
	}
	// 遍历数组 把剩下'O'转为'X'（走过一遍后没被包围的'O'）都变成'F'了 需要把F还原
	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			if board[i][j] == 'F' {
				board[i][j] = 'O'
			} else {
				board[i][j] = 'X'
			}
		}
	}

}

func dfs1(board [][]byte, i, j int) {
	m := len(board)
	n := len(board[0])
	if i < 0 || j < 0 || i >= m || j >= n || board[i][j] != 'O' {
		return
	}
	//如果是'O' 感染为'F'
	board[i][j] = 'F'
	//上下左右四个方向找寻相连的'O'
	dfs1(board, i-1, j)
	dfs1(board, i+1, j)
	dfs1(board, i, j-1)
	dfs(board, i, j+1)
}
